### PR TITLE
Core: Update ResolvingFileIO

### PR DESCRIPTION
**Pull request by rdblue**
This includes changes that I thought had been pushed in #3593 to fix Dan's review comments.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java [hunk 1]
```diff
@@ -120,15 +118,20 @@ private FileIO io(String location) {
...
-        } catch (IllegalArgumentException suppressed) {
-          LOG.warn("Failed to load FileIO implementation: {} (fallback)", FALLBACK_IMPL, suppressed);
-          // both attempts failed, throw the original exception with the later exception suppressed
-          e.addSuppressed(suppressed);
+        if (impl.equals(FALLBACK_IMPL)) {
+          // no implementation to fall back to, throw the exception
           throw e;
+        } else {
+          // couldn't load the normal class, fall back to HadoopFileIO
+          LOG.warn("Failed to load FileIO implementation: {}, falling back to {}", impl, FALLBACK_IMPL, e);
```
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java:126]: Nit: Maybe `... falling back to only using ${FALLBACK_IMPL}`?

Something to really clarify that this is the only implementation that will be used. Up to you if you feel it's clear enough as is (probably it is).

**Review by kbendick — APPROVED** [review]
One small nit but overall this looks good. Thanks @rdblue!

**Review by danielcweeks — APPROVED** [review]

File: core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java [hunk 2]
```diff
@@ -120,15 +118,20 @@ private FileIO io(String location) {
...
+          try {
+            io = CatalogUtil.loadFileIO(FALLBACK_IMPL, properties, conf);
+          } catch (IllegalArgumentException suppressed) {
+            LOG.warn("Failed to load FileIO implementation: {} (fallback)", FALLBACK_IMPL, suppressed);
+            // both attempts failed, throw the original exception with the later exception suppressed
+            e.addSuppressed(suppressed);
+            throw e;
+          }
```
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java:134]: Does catching and performing work in here generate any warnings?

I _think_ that's only for returning, but checking.

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java:143]: I'm not quite sure what you mean

**Comment by rdblue** [comment]
Thanks for bumping this PR, @kbendick! I didn't realize it was still outstanding.

**Pull request merged**